RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc

```diff
@@ -222,6 +222,11 @@ void PixelThresholdClusterizer::copy_to_buffer(DigiIterator begin, DigiIterator
     // std::cout << (doMissCalibrate ? "VI from db" : "VI linear") << std::endl;
   }
 #endif
+
+  //avoid undefined behavior
+  if (end <= begin)
```

Comment:
wouldn't it be better to essentially assert/crash here, and ensure that copy_to_buffer is not called with incorrect inputs?

Reply:
I don't personally have a strong preference on this. But I am not an expert in the clusterizer code so someone else would have to try to understand why `copy_to_buffer` is being called with these incorrect inputs and come up with a fix (maybe @ferencek or @czangela ?)

Reply:
No strong preference here either but calling `copy_to_buffer` in https://github.com/cms-sw/cmssw/blob/CMSSW_12_1_0_pre3/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc#L151 only if `end>begin` perhaps would be a more elegant solution.

Reply:
On the other hand, I think it makes sense for a function itself to act in case it encounters an undefined behavior rather than to leave the checking to the caller. So in the end I think I like the current fix more.

In general, I am not a big fan of asserts in the production code. For anything unexpected or undefined, isn't it better to deal with it gracefully and report a LogError? In this particular case it looks like we are encountering a situation where a pixel module has zero digis produced and an empty vector (DetSet) is passed on to the clustering routine. So this I would say is nothing particularly alarming and probably does not even require issuing a LogError. The following commented out line https://github.com/cms-sw/cmssw/blob/CMSSW_12_1_0_pre3/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc#L135 seems to suggest that this scenario can indeed occur. However, what I am a bit confused about is why these empty DetSets are not simply dropped from the digi collection? Either way, the clusterizer code should be able to handle such cases without any trouble.

Reply:
is this empty case corresponding to `(end == begin)` ?
or is this a case of end before begin?

Reply:
@slava77, `end == begin`

@OzAmram, I had a quick chat with @tsusa today precisely about this issue of `end == begin` which suggests that in the digi collection we have an empty vector of digis stored for a particular detId. The question then is why is this empty vector not dropped from the collection in the first place. It would therefore be good to check how that happens. On the other hand, even if there is some issue in the digi producer which can lead to such situations, the clusterizer should be immune against such cases which is what this PR achieves.

Reply:
Yeah I tend to agree that it would be good if the clusterizer does not fail for such a case. Maybe we add `LogWarning` message to this PR and followup later to try and track down the upstream issue?

Reply:
In my view, at least the code comments should be reasonably clear why a function is expected to sometimes get incorrect inputs. Something like "avoid undefined behaviour" may be quite mysterious for a reader later.

So if fixing this at the source is out of scope, how about:
```C++
  //In rare cases, this function gets called with an empty DetSet. This is not expected to be a problem because of XYZ
  if (end <= begin)
    return;
```

Reply:
`begin == end` is a reasonable case for a generic caller and it makes sense to me that if there is some preamble computation in the method to skip it if it's clearly not needed.
However, the part with `end < begin` looks bad and better be resolved at the caller.

Reply:
I agree with @slava77. The `end < begin` case looks bad but by construction it is not possible in this particular case. In the caller code the `begin` and `end` are iterators from the same vector so in the worst case `end==begin`.  So for such cases the method could issue a LogWarning and return.